README.md
cocor_compiler
==============

Use of Coco/R as practical exercise for Compiler Design Theory course

Antonio Nikolov 10378197 CS3071 Final Assignment Report
Abbreviations
codegen - Code Generator
symtab - Symbol Table
RT - runtime
CT - compile time
ATG - Tastier.ATG
EBNF - TastierEBNF.txt

Overview of bug fixes, modifications and additions made since last version:

added heap memory to the codegen to handle storage along with the stack.
fixed for loop so it does not ignore the first statement in the loop.
added switch statement that works with int.
reworked strings so they are handled on the heap.
fixed bug where global constants do not get initialized.
added any dimensional array structure with  RT initialization and RT bounds checking that can store int and bool.
added a “Record” structured data type similar to C structs in use that can contain int,bool,string and arrays and constants.
Now there is the ability to write code that is evaluated at RT in the global scope.This to allow for proper constant initialization and record declarations containing arrays. Normal (non-constant) variable and array can be declared in the global scope but not initialized.

Conventions I use :

I refer to single variables int,bool,string as “var” or “variable” where as array as “arr“ or “array” even though array is also a variable.
Above each symtab,codegen procedures and ATG production you can see a comment with:
no initials - no changes made 
AN - small modifications of how the production/procedure works e.g. type.
AN+ - heavy modifications of how the an existing production/procedure works e.g Stat.
AN++ - productions/procedure which I have defined e.g. ArrayDecl.
The EBNF file is up-to-date so refer to it for a simple view of the ATG without the semantic processing
There is extra code printed out when tastier program is interpreted. It is wrapped around curly brackets on the console.


Brief guidelines for the use of the modified/added Tastier language features from a Tastier programmer perspective (refer to EBNF for actual syntax):

Constants: can be declared in any scope including global,local and record scope. Can be initialized with expressions of any type and cannot be changed once assigned. Note that they do not behave like the C style #define macro which just replaces code at compile time.
Switch statement : works only with int expressions.
String type: strings are assigned in the same style as ints and bools.
Arrays: works with int and bool type and can be of any dimensions less than 256 :)          (although did not test with more than four). Can be declared with any integer expression including previously assigned variables and constants in any scope. Example use: 
int i , j;  
i := 4; j := 3;
 array bool myarray[ i ] [ j ] [ i + 3];
myarray[1][2][2] := true;
write myarray[1][2][2];
Record: declared in the global scope only. can contain constants,arrays and any variable type. Variables from the record are accessed by <record_name> “.” <variable_name> and arrays by <record_name> “->” <array_name>.The original record can be used directly or it can be instantiated with another record name by “new” <record_name> <new_record_name>. The names of variables and arrays inside the record do not clash with other global and local names. Example use:
program Tastier{
     const index1 := 3;
     record myrec{
            const index2 := 3;
            int i,j,k;
            string name;
            array int myarr[index1][index2][5];
     }
…........
   void function(){
          int i,j;
          myrec.i := 10;
          i := 11;
         write myrec.i;
         write i + myrec.i;
         new myrec myrec2;
         write myrec2.i // should be print 0
         j := 3;
       write myrec->myarr[1][2][0];
    }


New and modified ATG productions(refer to Tastier.ATG for actual code):

Switch: the condition of the switch is an expression so it leave a value on top of the stack when evaluated. Lets call it “dummy” A new object is created for dummy. The value of dummy  is stored on the stack where the object adr is pointing to. Later, for each “case” expression dummy is loaded after the value of the expression. Equality check and conditional jump are emitted. The address of each conditional jump is saved in variable “caseadr”. At the end of each case statement:  caseadr is patched with the current pc; also if there is a break clause then an unconditional jump is emitted and its address is saved in a List. At the end of the switch statement the all the addresses in the List are patched with the pc that is now just after the switch statement.


VarDecl<int kind,string rec>: kind and rec are inherited attributes where kind tells whether the new object is to be a regular variable or a record variable. rec tells the name of the record it belongs to. Regular variables a part of the empty “” record.


ConstVarDecl<int kind,string rec>: same as above. Also the created object is immutable, and expression is produced and its value assigned(at RT).
ArrayDecl<int kind,string rec>: attribute description same as above except kind can be either regular array or record array. After the ArrayPart is produced a request is made to the codegen to emit an instruction that puts the current heap pointer on the stack. The pointer is stored so this array identified its ident name can be referenced later on. Then the codegen must produce an ALLOCARR instruction.


ArrayPart<out int dim> : synthesizes the number of dimensions. At the end of this production the values of the each dimension will be on the stack with the right most on top. 


Record: produces the complete declaration for a record.


RecordDecl: produces a new copy of the original record so the new copy can be used as a separate record with the same fields as the original.


RecordVarPart<Obj rec, out Obj var>: inherits rec, synthesizes var. Produces a variable ident and checks if it belongs to the inherited record and var syntesizes the value of the variable.


RecordArrPart<Obj rec, out Obj var>: same as above except for arrays.


Factor<out int type>: syntesizes type. Changed so it handles arrays and record fields properly. I feel there is no point for me to go over the selection set of the production.


Stat: same as above.


ProcDecl<out string name>: synthesizes name. No longer sets the progStart. progStart always starts with the first instruction.
Tastier: Changed so now inserts JMP instructions just before each procedure except Main

I have written comments in the codegen and symtab file where I felt there was a need for more thorough explanation of the added opcodes and procedures.



changes.txt
commit 1ebdb2e18432325126d6be10befc9d1aa1d55b8d
Author: Antonio Nikolov <nikolova@tcd.ie>
Date:   Wed Nov 28 13:25:49 2012 +0000

    AB orig

diff --git a/TastierEBNF.html b/TastierEBNF.html
index f4dabe9..a10d7c8 100755
--- a/TastierEBNF.html
+++ b/TastierEBNF.html
@@ -2,23 +2,22 @@ EBNF Grammar of Tastier
 
    
 Taste =      "program" ident "{" {VarDecl} {ProcDecl} "}".
-ConstVarDecl = "const" ident ":=" Expr ";". //AN
 VarDecl =    Type ident {"," ident} ";".
-ProcDecl =   "void" ident "(" ")" "{" {VarDecl | Stat | ProcDecl} "}". //AN
-Type =       "int" | "bool | string". //AN
+ProcDecl =   "void" ident "(" ")" "{" {VarDecl | Stat | ProcDecl} "}".
+Type =       "int" | "bool".
 
-Stat =       ident (":=" Expr ";" | "(" ")" ";")
+Stat =       ident ("=" Expr ";" | "(" ")" ";")
            | "if" "(" Expr ")" Stat ["else" Stat]
            | "while "(" Expr ")" Stat
            | "read" ident ";"
-           | "write" Expr { "++" Expr} ";" //AN
-           | "{" {Stat | VarDecl | ConstVarDecl} "}". //AN
+           | "write" Expr ";"
+           | "{" {Stat | VarDecl} "}".
 
 Expr =       SimExpr [RelOp SimExpr].
 SimExpr =    Term {AddOp Term}.
 Term =       Factor {MulOp Factor}.
-Factor =     (ident | "true" | "false" | number | "-" Factor | string). //AN
+Factor =     (ident | "true" | "false" | number | "-" Factor).
 
 AddOp =      "+" | "-".
 MulOp =      "*" | "/".
-RelOp =      "=" | "<" | ">" | "!=" | "<=" | ">=". //AN
+RelOp =      "==" | "<" | ">".
\ No newline at end of file

commit c5680baba4782641d1e021805bdcbfc3c9f8e14e
Author: Antonio Nikolov <nikolova@tcd.ie>
Date:   Wed Nov 28 11:30:18 2012 +0000

    with strings

diff --git a/TastierEBNF.html b/TastierEBNF.html
new file mode 100755
index 0000000..f4dabe9
--- /dev/null
+++ b/TastierEBNF.html
@@ -0,0 +1,24 @@
+EBNF Grammar of Tastier
+
+   
+Taste =      "program" ident "{" {VarDecl} {ProcDecl} "}".
+ConstVarDecl = "const" ident ":=" Expr ";". //AN
+VarDecl =    Type ident {"," ident} ";".
+ProcDecl =   "void" ident "(" ")" "{" {VarDecl | Stat | ProcDecl} "}". //AN
+Type =       "int" | "bool | string". //AN
+
+Stat =       ident (":=" Expr ";" | "(" ")" ";")
+           | "if" "(" Expr ")" Stat ["else" Stat]
+           | "while "(" Expr ")" Stat
+           | "read" ident ";"
+           | "write" Expr { "++" Expr} ";" //AN
+           | "{" {Stat | VarDecl | ConstVarDecl} "}". //AN
+
+Expr =       SimExpr [RelOp SimExpr].
+SimExpr =    Term {AddOp Term}.
+Term =       Factor {MulOp Factor}.
+Factor =     (ident | "true" | "false" | number | "-" Factor | string). //AN
+
+AddOp =      "+" | "-".
+MulOp =      "*" | "/".
+RelOp =      "=" | "<" | ">" | "!=" | "<=" | ">=". //AN


SymTab.cs
using System;

//Antonio Nikolov 10378197 Tutorial 4
namespace Tastier {

public class Obj {  // object describing a declared name
	public string name;		// name of the object
	public int type;			// type of the object (undef for proc)
	public Obj	next;			// to next object in same scope
	public int kind;      // var, proc, scope
	public int adr;				// address in memory or start of proc
	public int level;			// nesting level; 0=global, 1=local
	public Obj locals;		// scopes: to locally declared objects
	public int nextAdr;		// scopes: next free address in this scope
	public int mutability; //AN constant or non-constant
}

public class SymbolTable {
	string [] types = new string [4]  {"Undef","Integer","Bool","String"};
	string [] kinds = new string [3] {"Var" ,"Proc" ,"Scope"};
	string [] levels = new string [2] {"Global" ," Local"};
	string [] is_mutable= new string [2] {"Mutable" ," Constant"};

	const int // types
		undef = 0, integer = 1, boolean = 2, str = 3;

	const int // object kinds
		var = 0, proc = 1, scope = 2;
	const int
		immutable = 1 , mutable = 0;


	public int curLevel;	// nesting level of current scope
	public Obj undefObj;	// object node for erroneous symbols
	public Obj topScope;	// topmost procedure scope

	Parser parser;
	
	public SymbolTable(Parser parser) {
		this.parser = parser;
		topScope = null;
		curLevel = -1;
		undefObj = new Obj();
		undefObj.name  =  "undef"; undefObj.type = undef; undefObj.kind = var;
		undefObj.adr = 0; undefObj.level = 0; undefObj.next = null;
	}


	// open a new scope and make it the current scope (topScope)
	public void OpenScope (string name) {
		Obj scop = new Obj();
		scop.name = name ; scop.kind = scope; 
		scop.locals = null; scop.nextAdr = 0;
		scop.next = topScope; topScope = scop; 
		curLevel++;
	
	}


	public void CloseScope () {
		//Console.WriteLine("Closing the Scope of :" + topScope.name); // AN
		Obj temp = topScope.locals;
		topScope = topScope.next; curLevel--;

		//-------------------------------------------AN------------------------------
		while(temp != null){
			Console.WriteLine( temp.name + 
						"\t| " + kinds[temp.kind] + 
						"\t| " + types[temp.type] +  
						"\t| " + levels[temp.level] +
						"\t| " + is_mutable[temp.mutability] + 
						"\t| " + temp.adr);
			temp = temp.next;
		}
		//END-------------------------------------------AN------------------------------
	}
	
	// create a new object node in the current scope
	public Obj NewObj (string name, int kind, int type) {
		Obj p, last, obj = new Obj();
		obj.name = name; obj.kind = kind; obj.type = type;
		obj.level = curLevel;
		p = topScope.locals; last = null;
		while (p != null) { 
			if (p.name == name) parser.SemErr("name declared twice");
			last = p; p = p.next;
		}
		if (last == null) topScope.locals = obj; else last.next = obj;
		if (kind == var) {
		    obj.adr = topScope.nextAdr++;
		    obj.mutability = mutable; //assume variables are not constant by default
		}
	      return obj;
	}

	//-------------------------------------------AN------------------------------
	// left the NewObj procedure as it is so I dont have to change the code elsewhere which calls NewObj
	// make new object with explicit const/ non-const
	//could have been generalized as well to  account for constant procedures in case of future extension i.e ones that cannot be overlapped
	public Obj NewConstVar (string name){
		Obj obj;
		obj = NewObj (name,var,undef);
		obj.mutability = immutable;
		return obj;
	}
	//assign the type of constant variable
	public void assignType (string name, int type){
		Obj obj = Find(name); 
		if( obj.mutability != immutable) parser.SemErr("cannot reasign type of non-constant variable");
		obj.type = type;
	}
	//END-------------------------------------------AN------------------------------
	
	// search the name in all open scopes and return its object node
	public Obj Find (string name) {
		Obj obj, scope;
		scope = topScope;
		while (scope != null) {  // for all open scopes
			obj = scope.locals;
			while (obj != null) {  // for all objects in this scope
				if (obj.name == name) return obj;
				obj = obj.next;
			}
			scope = scope.next;
		}
		parser.SemErr(name + " is undeclared");
		return undefObj;
	}

} // end SymbolTable

} // end namespace


Parser.cs

using System;

namespace Tastier {



public class Parser {
	public const int _EOF = 0;
	public const int _ident = 1;
	public const int _number = 2;
	public const int _string = 3;
	public const int maxT = 34;

	const bool T = true;
	const bool x = false;
	const int minErrDist = 2;
	
	public Scanner scanner;
	public Errors  errors;

	public Token t;    // last recognized token
	public Token la;   // lookahead token
	int errDist = minErrDist;

const int // types
      undef = 0, integer = 1, boolean = 2,str = 3;

   const int // object kinds
      var = 0, proc = 1;
  const int 
      mutable = 0, immutable = 1;

   public SymbolTable   tab;
   public CodeGenerator gen;
  
/*--------------------------------------------------------------------------*/


	public Parser(Scanner scanner) {
		this.scanner = scanner;
		errors = new Errors();
	}

	void SynErr (int n) {
		if (errDist >= minErrDist) errors.SynErr(la.line, la.col, n);
		errDist = 0;
	}

	public void SemErr (string msg) {
		if (errDist >= minErrDist) errors.SemErr(t.line, t.col, msg);
		errDist = 0;
	}
	
	void Get () {
		for (;;) {
			t = la;
			la = scanner.Scan();
			if (la.kind <= maxT) { ++errDist; break; }

			la = t;
		}
	}
	
	void Expect (int n) {
		if (la.kind==n) Get(); else { SynErr(n); }
	}
	
	bool StartOf (int s) {
		return set[s, la.kind];
	}
	
	void ExpectWeak (int n, int follow) {
		if (la.kind == n) Get();
		else {
			SynErr(n);
			while (!StartOf(follow)) Get();
		}
	}


	bool WeakSeparator(int n, int syFol, int repFol) {
		int kind = la.kind;
		if (kind == n) {Get(); return true;}
		else if (StartOf(repFol)) {return false;}
		else {
			SynErr(n);
			while (!(set[syFol, kind] || set[repFol, kind] || set[0, kind])) {
				Get();
				kind = la.kind;
			}
			return StartOf(syFol);
		}
	}

	
	void AddOp(out Op op) {
		op = Op.ADD; 
		if (la.kind == 4) {
			Get();
		} else if (la.kind == 5) {
			Get();
			op = Op.SUB; 
		} else SynErr(35);
	}

	void ConstVarDecl() {
		string name; int type; Obj obj; 
		Expect(6);
		Ident(out name);
		obj = tab.NewConstVar(name);  
		Expect(7);
		Expr(out type);
		tab.assignType(name,type); 
		if (obj.level == 0) gen.Emit(Op.STOG, obj.adr);
		else gen.Emit(Op.STO, tab.curLevel-obj.level, obj.adr); 
		Expect(8);
	}

	void Ident(out string name) {
		Expect(1);
		name = t.val; 
	}

	void Expr(out int type) {
		int type1; Op op; 
		SimExpr(out type);
		if (StartOf(1)) {
			RelOp(out op);
			SimExpr(out type1);
			if (type != type1) SemErr("incompatible types");
			gen.Emit(op); type = boolean; 
		}
	}

	void SimExpr(out int type) {
		int type1; Op op; 
		Term(out type);
		while (la.kind == 4 || la.kind == 5) {
			AddOp(out op);
			Term(out type1);
			if (type != integer || type1 != integer) 
			  SemErr("integer type expected");
			gen.Emit(op); 
		}
	}

	void RelOp(out Op op) {
		op = Op.EQU; 
		switch (la.kind) {
		case 18: {
			Get();
			break;
		}
		case 19: {
			Get();
			op = Op.LSS; 
			break;
		}
		case 20: {
			Get();
			op = Op.GTR; 
			break;
		}
		case 21: {
			Get();
			op = Op.NEQ; 
			break;
		}
		case 22: {
			Get();
			op = Op.LSE; 
			break;
		}
		case 23: {
			Get();
			op = Op.GTE; 
			break;
		}
		default: SynErr(36); break;
		}
	}

	void Factor(out int type) {
		int n; Obj obj; string name; string s; 
		type = undef; 
		switch (la.kind) {
		case 1: {
			Ident(out name);
			obj = tab.Find(name); type = obj.type;
			if (obj.kind == var) {
			  if (obj.level == 0) gen.Emit(Op.LOADG, obj.adr);
			//                                else gen.Emit(Op.LOAD, obj.adr);                            ***
			     else gen.Emit(Op.LOAD, tab.curLevel-obj.level, obj.adr); // 
			} else SemErr("variable expected"); 
			break;
		}
		case 2: {
			Get();
			n = Convert.ToInt32(t.val); 
			gen.Emit(Op.CONST, n); type = integer; 
			break;
		}
		case 5: {
			Get();
			Factor(out type);
			if (type != integer) {
			  SemErr("integer type expected"); type = integer;
			}
			gen.Emit(Op.NEG); 
			break;
		}
		case 9: {
			Get();
			gen.Emit(Op.CONST, 1); type = boolean; 
			break;
		}
		case 10: {
			Get();
			gen.Emit(Op.CONST, 0); type = boolean; 
			break;
		}
		case 3: {
			Get();
			s = t.val; gen.EmitStr(s); type = str; 
			break;
		}
		default: SynErr(37); break;
		}
	}

	void MulOp(out Op op) {
		op = Op.MUL; 
		if (la.kind == 11) {
			Get();
		} else if (la.kind == 12) {
			Get();
			op = Op.DIV; 
		} else SynErr(38);
	}

	void ProcDecl() {
		string name; Obj obj; int adr, adr2; 
		Expect(13);
		Ident(out name);
		obj = tab.NewObj(name, proc, undef); obj.adr = gen.pc;
		//                          if (name == "Main") gen.progStart = gen.pc;        
		if (name == "Main") {                           // 
		  obj.level= 0; gen.progStart = gen.pc;        // 
		}                                               // 
		  else obj.level = tab.curLevel+1;             // 
		tab.OpenScope(name); 
		Expect(14);
		Expect(15);
		Expect(16);
		gen.Emit(Op.ENTER, 0); adr = gen.pc - 2; 
		while (StartOf(2)) {
			if (la.kind == 6) {
				ConstVarDecl();
			} else if (la.kind == 31 || la.kind == 32 || la.kind == 33) {
				VarDecl();
			} else if (StartOf(3)) {
				Stat();
			} else {
				gen.Emit(Op.JMP, 0); adr2 = gen.pc - 2; 
				ProcDecl();
				gen.Patch(adr2, gen.pc); 
			}
		}
		Expect(17);
		gen.Emit(Op.LEAVE); gen.Emit(Op.RET);
		gen.Patch(adr, tab.topScope.nextAdr);
		tab.CloseScope(); 
	}

	void VarDecl() {
		string name; int type; 
		Type(out type);
		Ident(out name);
		tab.NewObj(name, var, type); 
		while (la.kind == 29) {
			Get();
			Ident(out name);
			tab.NewObj(name, var, type); 
		}
		Expect(8);
	}

	void Stat() {
		int type; string name; Obj obj;
		int adr, adr2, loopstart; 
		switch (la.kind) {
		case 1: {
			Ident(out name);
			obj = tab.Find(name); 
			if (la.kind == 7) {
				Get();
				if (obj.kind != var) SemErr("cannot assign to procedure");	//AN
				if (obj.mutability == immutable) SemErr("cannot reassign constant variable"); 
				Expr(out type);
				Expect(8);
				if (type != obj.type) SemErr("incompatible types");
				if (obj.level == 0) gen.Emit(Op.STOG, obj.adr);
				  else gen.Emit(Op.STO, tab.curLevel-obj.level, obj.adr); 
			} else if (la.kind == 14) {
				Get();
				Expect(15);
				Expect(8);
				if (obj.kind != proc) SemErr("object is not a procedure");
				  gen.Emit(Op.CALL, obj.level-tab.curLevel, obj.adr); 
			} else SynErr(39);
			break;
		}
		case 24: {
			Get();
			Expect(14);
			Expr(out type);
			Expect(15);
			if (type != boolean) SemErr("boolean type expected");
			  gen.Emit(Op.FJMP, 0); adr = gen.pc - 2; 
			Stat();
			if (la.kind == 25) {
				Get();
				gen.Emit(Op.JMP, 0); adr2 = gen.pc - 2;
				gen.Patch(adr, gen.pc); adr = adr2; 
				Stat();
			}
			gen.Patch(adr, gen.pc); 
			break;
		}
		case 26: {
			Get();
			loopstart = gen.pc; 
			Expect(14);
			Expr(out type);
			Expect(15);
			if (type != boolean) SemErr("boolean type expected");
			  gen.Emit(Op.FJMP, 0); adr = gen.pc - 2; 
			Stat();
			gen.Emit(Op.JMP, loopstart); gen.Patch(adr, gen.pc); 
			break;
		}
		case 27: {
			Get();
			Ident(out name);
			Expect(8);
			obj = tab.Find(name);
			if (obj.type != integer) SemErr("integer type expected");
			  gen.Emit(Op.READ);
			if (obj.level == 0) gen.Emit(Op.STOG, obj.adr);
			//                             else gen.Emit(Op.STO, obj.adr); .)                            ***
			  else gen.Emit(Op.STO, tab.curLevel-obj.level, obj.adr); 
			break;
		}
		case 28: {
			Get();
			Expr(out type);
			if(type == str) gen.Emit(Op.SWRITE);					//AN
			    else if(type == integer) gen.Emit(Op.WRITE);		//AN
			else SemErr("expected type int or string"); 
			while (la.kind == 29) {
				Get();
				Expr(out type);
				if(type == str) gen.Emit(Op.SWRITE); 					//AN
				     else if(type == integer) gen.Emit(Op.WRITE);		//AN
				else SemErr("expected type int or string"); 
			}
			Expect(8);
			gen.Emit(Op.NEWLINE);
			break;
		}
		case 16: {
			Get();
			while (StartOf(4)) {
				if (StartOf(3)) {
					Stat();
				} else if (la.kind == 31 || la.kind == 32 || la.kind == 33) {
					VarDecl();
				} else {
					ConstVarDecl();
				}
			}
			Expect(17);
			break;
		}
		default: SynErr(40); break;
		}
	}

	void Term(out int type) {
		int type1; Op op; 
		Factor(out type);
		while (la.kind == 11 || la.kind == 12) {
			MulOp(out op);
			Factor(out type1);
			if (type != integer || type1 != integer) 
			  SemErr("integer type expected");
			gen.Emit(op); 
		}
	}

	void Tastier() {
		string name; 
		Expect(30);
		Ident(out name);
		tab.OpenScope(name); 
		Expect(16);
		while (StartOf(5)) {
			if (la.kind == 6) {
				ConstVarDecl();
			} else if (la.kind == 31 || la.kind == 32 || la.kind == 33) {
				VarDecl();
			} else {
				ProcDecl();
			}
		}
		Expect(17);
		tab.CloseScope();
		if (gen.progStart == -1) SemErr("main function never defined");
		
	}

	void Type(out int type) {
		type = undef; 
		if (la.kind == 31) {
			Get();
			type = integer; 
		} else if (la.kind == 32) {
			Get();
			type = boolean; 
		} else if (la.kind == 33) {
			Get();
			type = str; 
		} else SynErr(41);
	}



	public void Parse() {
		la = new Token();
		la.val = "";		
		Get();
		Tastier();
		Expect(0);

	}
	
	static readonly bool[,] set = {
		{T,x,x,x, x,x,x,x, x,x,x,x, x,x,x,x, x,x,x,x, x,x,x,x, x,x,x,x, x,x,x,x, x,x,x,x},
		{x,x,x,x, x,x,x,x, x,x,x,x, x,x,x,x, x,x,T,T, T,T,T,T, x,x,x,x, x,x,x,x, x,x,x,x},
		{x,T,x,x, x,x,T,x, x,x,x,x, x,T,x,x, T,x,x,x, x,x,x,x, T,x,T,T, T,x,x,T, T,T,x,x},
		{x,T,x,x, x,x,x,x, x,x,x,x, x,x,x,x, T,x,x,x, x,x,x,x, T,x,T,T, T,x,x,x, x,x,x,x},
		{x,T,x,x, x,x,T,x, x,x,x,x, x,x,x,x, T,x,x,x, x,x,x,x, T,x,T,T, T,x,x,T, T,T,x,x},
		{x,x,x,x, x,x,T,x, x,x,x,x, x,T,x,x, x,x,x,x, x,x,x,x, x,x,x,x, x,x,x,T, T,T,x,x}

	};
} // end Parser


public class Errors {
	public int count = 0;                                    // number of errors detected
	public System.IO.TextWriter errorStream = Console.Out;   // error messages go to this stream
	public string errMsgFormat = "-- line {0} col {1}: {2}"; // 0=line, 1=column, 2=text

	public virtual void SynErr (int line, int col, int n) {
		string s;
		switch (n) {
			case 0: s = "EOF expected"; break;
			case 1: s = "ident expected"; break;
			case 2: s = "number expected"; break;
			case 3: s = "string expected"; break;
			case 4: s = "\"+\" expected"; break;
			case 5: s = "\"-\" expected"; break;
			case 6: s = "\"const\" expected"; break;
			case 7: s = "\":=\" expected"; break;
			case 8: s = "\";\" expected"; break;
			case 9: s = "\"true\" expected"; break;
			case 10: s = "\"false\" expected"; break;
			case 11: s = "\"*\" expected"; break;
			case 12: s = "\"/\" expected"; break;
			case 13: s = "\"void\" expected"; break;
			case 14: s = "\"(\" expected"; break;
			case 15: s = "\")\" expected"; break;
			case 16: s = "\"{\" expected"; break;
			case 17: s = "\"}\" expected"; break;
			case 18: s = "\"=\" expected"; break;
			case 19: s = "\"<\" expected"; break;
			case 20: s = "\">\" expected"; break;
			case 21: s = "\"!=\" expected"; break;
			case 22: s = "\"<=\" expected"; break;
			case 23: s = "\">=\" expected"; break;
			case 24: s = "\"if\" expected"; break;
			case 25: s = "\"else\" expected"; break;
			case 26: s = "\"while\" expected"; break;
			case 27: s = "\"read\" expected"; break;
			case 28: s = "\"write\" expected"; break;
			case 29: s = "\",\" expected"; break;
			case 30: s = "\"program\" expected"; break;
			case 31: s = "\"int\" expected"; break;
			case 32: s = "\"bool\" expected"; break;
			case 33: s = "\"string\" expected"; break;
			case 34: s = "??? expected"; break;
			case 35: s = "invalid AddOp"; break;
			case 36: s = "invalid RelOp"; break;
			case 37: s = "invalid Factor"; break;
			case 38: s = "invalid MulOp"; break;
			case 39: s = "invalid Stat"; break;
			case 40: s = "invalid Stat"; break;
			case 41: s = "invalid Type"; break;

			default: s = "error " + n; break;
		}
		errorStream.WriteLine(errMsgFormat, line, col, s);
		count++;
	}

	public virtual void SemErr (int line, int col, string s) {
		errorStream.WriteLine(errMsgFormat, line, col, s);
		count++;
	}
	
	public virtual void SemErr (string s) {
		errorStream.WriteLine(s);
		count++;
	}
	
	public virtual void Warning (int line, int col, string s) {
		errorStream.WriteLine(errMsgFormat, line, col, s);
	}
	
	public virtual void Warning(string s) {
		errorStream.WriteLine(s);
	}
} // Errors


public class FatalError: Exception {
	public FatalError(string m): base(m) {}
}
}